cou/steps/plan.py

```diff
@@ -52,24 +53,163 @@
 logger = logging.getLogger(__name__)
 
 
+def pre_plan_sane_checks(upgrade_group: Optional[str], analysis_result: Analysis) -> None:
+    """Pre checks to generate the upgrade plan.
+
+    :param upgrade_group: Upgrade group passed by the user using the cli.
+    :type upgrade_group: Optional[str]
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    """
+    checks = [
+        is_valid_openstack_cloud,
+        is_supported_series,
+        is_highest_release_achieved,
+        is_target_supported,
+    ]
+    if upgrade_group == "data-plane":
+        checks.append(is_data_plane_ready_to_upgrade)
+    for check in checks:
+        check(analysis_result)
+
+
+def is_valid_openstack_cloud(analysis_result: Analysis) -> None:
+    """Check if the model passed is a valid OpenStack cloud.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot determine the current OS release
+        or Ubuntu series.
+    """
+    current_os_release = analysis_result.current_cloud_os_release
+    current_series = analysis_result.current_cloud_series
+    if not current_os_release:
+        raise NoTargetError(
+            "Cannot determine the current OS release in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+    if not current_series:
+        raise NoTargetError(
+            "Cannot determine the current Ubuntu series in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+
+def is_supported_series(analysis_result: Analysis) -> None:
+    """Check the Ubuntu series of the cloud to see if it is supported.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises OutOfSupportRange: When series is not supported.
+    """
+    current_series = analysis_result.current_cloud_series
+    supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
+    # series already checked at is_valid_openstack_cloud
+    if current_series not in supporting_lts_series:  # type: ignore
```

Comment:
I think it's cleaner to add `if current_series and current_series not in ...`, this check should not raise an exception if `current_series` is `None`.

Reply:
same for all other `# type: ignore`